### PR TITLE
Dual writes for users

### PR DESCRIPTION
**Pull request by rahilb**

### CODE REVIEW

File: app/models/database/mongo/PersistedUser.scala [hunk 1]
```diff
@@ -195,7 +195,7 @@ case class IdentityUser(
 object IdentityUser {
   val identityUserReads: Reads[IdentityUser] = (
       (JsPath \ "primaryEmailAddress").read[String] and
-      (JsPath \ "_id").read[String] and
+      (JsPath \ "_id").read[String].orElse((JsPath \ "id").read[String]) and
```
Comment (QuarpT) [inline app/models/database/mongo/PersistedUser.scala]: This should just be "id" after the upserts, unless it's inconsistent in mongo

Comment (rahilb) [reply app/models/database/mongo/PersistedUser.scala]: but we need to support stuff written by postgres too?

File: app/models/database/postgres/PostgresUserRepository.scala [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  with PostgresJsonFormats
+  with PostgresUtils {
+
+  def search(query: String, limit: Option[Int] = None, offset: Option[Int] = None): ApiResponse[SearchResponse] = {
+    val _offset = offset.getOrElse(0)
+    val _limit = limit.getOrElse(SearchValidation.maximumLimit)
+    val lowcaseQuery = query.toLowerCase
+    val sql =
+      sql"""
+       |SELECT jdoc, count(*) OVER() AS full_count FROM users
```
Comment (QuarpT) [inline app/models/database/postgres/PostgresUserRepository.scala]: I guess we have indexes on all the searchFields and private fields?

File: app/models/database/postgres/PostgresUserRepository.scala [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import scala.concurrent.{ExecutionContext, Future}
+import scalaz.-\/
+import scalaz.syntax.either._
+
+class PostgresUserRepository @Inject()(implicit ec: ExecutionContext) extends Logging
+  with PostgresJsonFormats
+  with PostgresUtils {
+
+  def search(query: String, limit: Option[Int] = None, offset: Option[Int] = None): ApiResponse[SearchResponse] = {
+    val _offset = offset.getOrElse(0)
```
Comment (QuarpT) [inline app/models/database/postgres/PostgresUserRepository.scala]: A large offset will result in a long running query

File: app/models/database/postgres/PostgresUserRepository.scala [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+       |OR jdoc #>> '{searchFields,displayName}' = ${lowcaseQuery}
+       |OR jdoc #>> '{privateFields,registrationIp}' = ${lowcaseQuery}
+       |OR jdoc #>> '{privateFields,lastActiveIpAddress}' = ${lowcaseQuery}
+       |LIMIT ${_limit}
+       |OFFSET ${_offset}
+       """.stripMargin
+    readOnly { implicit session =>
+      val results = sql.map(rs => Json.parse(rs.string(1)).as[IdentityUser] -> rs.int(2)).list().apply()
+      val count = results.headOption.map(_._2).getOrElse(0)
+      SearchResponse.create(count, _offset, results.map(_._1))
```
Comment (QuarpT) [inline app/models/database/postgres/PostgresUserRepository.scala]: Ideally the dao layer shouldn't be creating `ApiResponse[SearchResponse]` types

File: app/models/database/postgres/PostgresUserRepository.scala (hunk 2 above)
Comment (rahilb) [reply app/models/database/postgres/PostgresUserRepository.scala]: `idx_user_search`

File: app/models/database/postgres/PostgresUserRepository.scala (hunk 4 above)
Comment (rahilb) [reply app/models/database/postgres/PostgresUserRepository.scala]: A lot of things ideally shouldn't happen...

File: app/models/database/postgres/PostgresUserRepository.scala [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+import models.database.mongo.{IdentityUser, PersistedUser}
+import play.api.libs.json.Json
+import scalikejdbc._
+import spray.json.JsObject
+
+import scala.concurrent.{ExecutionContext, Future}
+import scalaz.-\/
+import scalaz.syntax.either._
+
+class PostgresUserRepository @Inject()(implicit ec: ExecutionContext) extends Logging
```
Comment (QuarpT) [inline app/models/database/postgres/PostgresUserRepository.scala]: Play default execution context is bad for blocking operations

**Comment by rahilb** [comment]
>Looks good, should set a maximum offset for users search since that query is potentially dangerous

Done: 187fc6b

**Pull request merged**